Bearded.Utilities/Algorithms/BinPacking.cs

```diff
@@ -173,56 +172,60 @@ private class Tree<T>
 
             private class Node
             {
-                private Node down;
-                private Node right;
+                private (Node Child1, Node Child2)? split;
 
                 public int X { get; }
                 public int Y { get; }
                 public int W { get; }
                 public int H { get; }
 
-                public Node(int x, int y, int w, int h, Node down, Node right)
+                public Node(int x, int y, int w, int h, (Node, Node)? split = null)
                 {
-                    this.down = down;
-                    this.right = right;
+                    this.split = split;
                     X = x;
                     Y = y;
                     W = w;
                     H = h;
                 }
 
-                public Node(int x, int y, int w, int h)
+                public Node? FindNodeWithSpaceFor(int w, int h)
                 {
-                    X = x;
-                    Y = y;
-                    W = w;
-                    H = h;
-                }
-
-                public Node Find(int w, int h)
-                {
-                    if (right != null)
-                        return right.Find(w, h) ?? down.Find(w, h);
+                    if (split is var (c1, c2))
+                        return c1.FindNodeWithSpaceFor(w, h) ?? c2.FindNodeWithSpaceFor(w, h);
                     if (w <= W && h <= H)
                         return this;
                     return null;
                 }
 
                 public void Split(int w, int h)
                 {
-                    down = new Node(X, Y + h, W, H - h);
-                    right = new Node(X + w, Y, W - w, h);
+                    /*
+                     Splitting a node semantically means reserving a (w, h) area and
+                     divvying up the remaining space as empty nodes as indicated in this diagram.
+
+                    X,Y-----+------+
+                     |   w  |      |
+                     |h     |child2|
+                     |      |      |
+                     +------+------+
+                     |    child1   |
+                     +------+------+
+                     */
+                    split = (
+                        new Node(X, Y + h, W, H - h),
+                        new Node(X + w, Y, W - w, h)
+                        );
                 }
 
-                public int GetEmptyPixels()
+                public int CountEmptyPixels()
                 {
-                    if (right == null)
-                        return W * H;
-                    return down.GetEmptyPixels() + right.GetEmptyPixels();
+                    if (split is var (down, right))
+                        return down.CountEmptyPixels() + right.CountEmptyPixels();
+                    return W * H;
                 }
             }
 
-            private Node root;
+            private Node root = null!;
```

Comment:
Optional:

Using `null!` is a bit of a hack IMO. I get why it is allowed here (it is used to share state between different methods called from `Fit`, and it makes sure it's always set), but I had to dig around in the code a bit. Should we adopt the standard where if we use `null!`, we must always add a comment explaining why it's acceptable in this case?

Reply:
I'm ok with introducing that rule, but in pretty much all cases we will do this because we're initialising somewhere other than the constructor, but the code still guarantees it will be initialised. Perhaps that in itself shouldn't be done. There may be better options like builder objects or smarter initial values, or...

Reply:
+1 that avoiding this situation is better than this hack. Forcing you to comment would force you to think about it too in a way.

Reply:
Found a way to avoid it, which is nice.